server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -345,9 +352,32 @@ private <C extends Collector, T> T search(Weight weight, CollectorManager<C, T>
 
     @Override
     public void search(List<LeafReaderContext> leaves, Weight weight, Collector collector) throws IOException {
-        collector.setWeight(weight);
-        for (LeafReaderContext ctx : leaves) { // search each subreader
-            searchLeaf(ctx, weight, collector);
+        try {
+            collector.setWeight(weight);
+            for (LeafReaderContext ctx : leaves) { // search each subreader
+                searchLeaf(ctx, weight, collector);
```

Comment:
Interestingly, Lucene seems to catch the error for each leaf, which means that time exceeded would be caught for all the following leaves? It seems to me that we are good catching it outside of the loop and terminate the collection at the first occurrence of the time exceeded exception.

Reply:
I think it is logical to stop iterating when we hit that exception, didn't know lucene is doing it per slice

Reply:
agreed